README.md
# Random People Generator

This package is just Scala-based command-line tool to create fake people
records, using first and last names taken at random from United
States Census Bureau data that's captured in local files. By default, it splits
the generated names so that half are female and half are male, but that can be
changed via command line options.

The tool can generate CSV or JSON output.

As is probably obvious, I use this program to generate test data.

# Installation

Clone this repo in the usual way. Then, read on.

`peoplegen` is built with [SBT](http://scala-sbt.org). You'll need to have
SBT installed to proceed. See <https://www.scala-sbt.org/download.html>
for details.

```
sbt install
```

will build a fat jar and install it in `$HOME/local/libexec`, by default.
It'll then install a wrapper `peoplegen` script in `$HOME/local/bin`. You
can change the prefix from `$HOME/local` to something else by setting the
`installDir` in `build.sbt` to a different path. See the commented out example
in `build.sbt`.

**Note for Windows users:** I don't run this thing on Windows, so I'm probably
not going to go out of my way to support it there. It *should* work fine, but
you're on your own if it doesn't.

# Usage

At any time, you can run `peoplegen --help` for a usage summary. The command
line looks like:

```
Usage: peoplegen [options] <total> [<outputfile>]
```

* `<total>` is the total number of people records to generate.
* `<outputfile>` is the file to which to write the records; if not supplied,
  the output goes to standard output.

## Options

`peoplegen` currently supports the following options:

* `--help`: Generate the usage message and exit.

* `-f <percent>` or `--female <percent>`: Percent of female records to
  generate.

* `-m <percent>` or `--male <percent>`: Percent of male records to generate.
  
**NOTE**: If you specify neither male nor female percentages, both default
to 50. If you specify only one percentage, the other is set to the remainder.
(e.g., If you specify only `--male 60`, the female percentage is set to 40.)
If you set _both_ percentages, they _must_ add up to 100, or `peoplegen`
will abort.

* `--id`: Generate unique per-row ID values.

* `--ssn`: Generate Social Security Number values. Note that the generated
  SSNs are deliberately invalid, as described at 
  <https://stackoverflow.com/a/2313726/53495>.
  
* `--salaries`: Generate salary data. Salaries are generated as a normal 
  distribution of integers, around a mean of 72,641 (the U.S. mean salary in 
  2014), with a sigma (i.e., a spread, or standard deviation) of 20,000. To 
  change these values, use --salary-mean and
                                 --salary-sigma.

* `--salary-mean <value>`: You can use this option to change the mean salary
  for the salary distribution. **Note**: Changing this value can result
  in negative salaries, so check your final data.

* `--salary-sigma <value>`: You can use this option to change the salary
  generation sigma—the spread, if you prefer. A smaller number means more
  salaries will cluster around the mean. A larger number means the distribution
  will be more "spread out". The distribution will still be a normal one (a
  bell curve), but the mean and the sigma control the _shape_ of the curve.

* `--year-min <value>`: Specify the starting year for birth dates. Defaults to 
  65 years ago from this year.
  
* `--year-max <value>`: Specify the ending year for birth dates. Defaults to 
  18 years ago from this year. This year cannot _precede_ the `year-min` value.

* `--delim <c>`: (CSV only) The delimiter to use between columns. The default
  is a comma (","). Any single character is fine. For tab, use the 2-character
  sequence "\t".

* `--header`: (CSV only) Generate a header for CSV output. Default: no header

* `-F <format>` or `--format <format>`: The file format to generate. Allowable
  values: "csv" or "json"

* `--camel`: Use camel case for CSV column names or JSON field names. For
  example: `firstName`, `lastName`

* `--english`: Use English (space-separated) names for column names. For
  example: `first name`, `last name`
  
* `--snake`: Use "snake case" (underscores) names for column names. For
  example: `first_name`, `last_name`

* `-j <format>` or `--json-format <format>`: (JSON only) Specify how the
  JSON should be generated. Legal values:
    * "rows" (default): generate individual rows of 1-line JSON people records.
      This format is useful with [Apache Spark](https://spark.apache.org).
    * "array": generate a JSON array with the JSON people records, all on
      one line        
    * "pretty": generate pretty-printed JSON.

* `-v` or `--verbose`: Emit (some) verbose processing messages.


CHANGELOG.md
# Change Log for _peoplegen_

**Version 2.1.4**

- Upgraded `sbt` version to 1.3.7.
- Got rid of Lightbend Activator.
- Upgraded to Scala 2.13 and fixed code that was using old APIs (e.g., now
  uses `LazyList` instead of `Stream`).
- Updated dependencies.

**Version 2.1.3**

Added verbose timings.

**Version 2.1.2**

Fixed bug in generation: Female names used for males.

**Version 2.1.1**

Fixed bug in generation of people: No males were being generated.

**Version 2.1.0**

Added `--id` option, allowing generation of unique per-row IDs.

**Version 2.0.0**

Ported from Python to Scala.


peoplegen/__init__.py
# -*- coding: utf-8 -*-
'''
peoplegen — Generate random people data.

Usage:
  peoplegen --help
  peoplegen [options] <total> <outputfile>

Options:
   -f PERCENT, --female PERCENT  Percentage of female names. Defaults to 50%.
   -m PERCENT, --male PERCENT    Percentage of male names. Defaults to 50%.
   --ssn                         Generate social security numbers.
   --header                      Generate header record if output is CSV.
   --salary RANGE                Generate salaries. Range is: LOWER-UPPER
   --yearmin YEAR                Minimum birth year. Defaults to 65 years ago.
   --yearmax YEAR                Maximum birth year. Defaults to 18 years ago.
   -s SEP, --sep SEP             Output field separator. [default: ,]

The output format is determined from the output file extension. Use '.json'
for JSON output and '.csv' for CSV output.

If you specify one of -m or -f, but not the other, the other will be calculated.
For instance, -f 25 means "25% females, 75% males". You can specify both.
'''

from __future__ import (absolute_import, division, print_function,
                        unicode_literals)

from builtins import (bytes, dict, int, list, object, range, str, ascii,
                      chr, hex, input, next, oct, open, pow, round, super,
                      filter, map, zip)
from future.standard_library import install_aliases
install_aliases()

import sys
import re
import docopt
import random
import os
from pkg_resources import resource_stream

__version__ = '2.0'
USAGE = __doc__

SSN_PREFIXES = list(range(900, 1000)) + [666]


def _load_data_file(name):
    here=os.path.dirname(os.path.abspath(__file__))
    files_to_check = (
        os.path.join(here, 'peoplegen', 'data', name),
        os.path.join(here, 'data, name')
    )

    res = None
    for path in files_to_check:
        if os.path.exists(path):
            with open(path) as f:
                res = [s.strip() for s in f.readlines()]
            break
    else:
        f = resource_stream('peoplegen.data', name)
        res = [s.strip() for s in f.readlines()]

    return res


def _load_data():
    last_names = _load_data_file('last_names.txt')
    male_first_names = _load_data_file('male_first_names.txt')
    female_first_names = _load_data_file('female_first_names.txt')
    return (last_names,
            [s.capitalize() for s in female_first_names],
            [s.capitalize() for s in male_first_names])


def _generate_salaries(total, low, high):
    # This is a hack to approximate a normal (bell curve) distribution while
    # eliminating values that fall outside the [lower, upper] range.
    mean = (low + high) / 2
    sigma = min((high - low) / 5, 30000)

    nums = []
    while len(nums) < total:
        n = int(random.normalvariate(mean, sigma))
        if (n >= low) and (n <= high):
            nums.append(n)

    return nums


def _make_fake_ssn():
    first = random.choice(SSN_PREFIXES)
    second = random.randint(10, 99)
    third = random.randint(1000, 9999)
    return '{0}-{1}-{2}'.format(first, second, third)


def _generate_name(first_names, last_names):
    first = random.choice(first_names)
    return (first,
            random.choice(first_names),
            random.choice(last_names))


def _generate_birth_date(min_year, max_year):
    from datetime import datetime, date
    import time
    import calendar

    today = datetime.today()
    if not min_year:
        min_year = today.year - 65
    if not max_year:
        max_year = today.year - 18

    min_date = date(year=min_year, month=1, day=1)
    max_date = date(year=max_year,month=12, day=31)

    # Convert to unix times.
    max_epoch = calendar.timegm(max_date.timetuple())
    min_epoch = calendar.timegm(min_date.timetuple())

    rand_epoch = random.randint(min_epoch, max_epoch)
    return unicode(time.strftime('%Y-%m-%d', time.localtime(rand_epoch)))

def _write_csv(file, delim, people, header, salaries=False, ssn=False):
    import csv
    import unicodedata
    sep = unicodedata.normalize('NFKD', delim).encode('ascii', 'ignore')

    with open(file, 'wb') as csvfile:
        w = csv.writer(csvfile, delimiter=sep)
        if header:
            headers = ['firstName', 'middleName', 'lastName', 'gender',
                       'birthDate']
            if salaries:
                headers.append('salary')
            if ssn:
                headers.append('ssn')
            w.writerow(headers)

        for person in people:
            data = [
                person['firstName'], person['middleName'], person['lastName'],
                person['gender'], person['birthDate']
            ]
            if salaries:
                data.append(person['salary'])
            if ssn:
                data.append(person['ssn'])
            w.writerow(data)

def _write_json(file, people):
    import json
    with open(file, 'wb') as out:
        json.dump(people, out)

def _die(msg):
    sys.stderr.write(msg + '\n')
    sys.exit(1)


def _die_usage(msg):
    _die(msg + '\n' + USAGE)


def _parse_salary(salary_range):
    range_pat = re.compile('^(\d+)-(\d+)$')
    try:
        m = range_pat.match(salary_range)
        if not m:
            _die_usage('Bad range for --salary: {0}'.format(salary_range))

        lower = int(m.group(1))
        upper = int(m.group(2))
        if lower >= upper:
            _die_usage('Empty range for --salary: {0}'.format(salary_range))
        return (lower, upper)
    except ValueError:
        _die(USAGE)


def _parse_percents(male_percent_str, female_percent_str):
    if not (male_percent_str or female_percent_str):
        # Neither specified. 50% each.
        male_percent = 50
        female_percent = 50
    else:
        r = re.compile('^\d+$')
        male_percent = None
        female_percent = None
        if male_percent_str:
            if not r.match(male_percent_str):
                _die_usage('Bad male percent: {0}'.format(male_percent_str))
            male_percent = int(male_percent_str)

        if female_percent_str:
            if not r.match(female_percent_str):
                _die_usage('Bad female percent: {0}'.format(female_percent_str))
            female_percent = int(female_percent_str)


        if not female_percent:
            # We have a male value, but not a female one.
            female_percent = 100 - male_percent
        if not male_percent:
            # We have a male value, but not a female one.
            male_percent = 100 - female_percent

    if (male_percent + female_percent) != 100:
        _die_usage("Male and female percentages don't add up to 100.")

    return (male_percent, female_percent)


def _parse_args():
    opts = docopt.docopt(USAGE, version=__version__)
    if opts['--salary']:
        opts['--salary'] = _parse_salary(opts['--salary'])
    (male_percent, female_percent) = _parse_percents(opts['--male'],
                                                     opts['--female'])
    opts['--male'] = male_percent
    opts['--female'] = female_percent
    opts['--sep'] = str(opts['--sep'])

    for key in ('--yearmin', '--yearmax', '<total>'):
        try:
            if opts[key]:
                opts[key] = int(opts[key])
        except ValueError:
            _die_usage('Bad value for {0}: "{1}"'.format(key, opts[key]))

    _, ext = os.path.splitext(opts['<outputfile>'])
    if ext == '.csv':
        opts['<format>'] = 'csv'
    elif ext == '.json':
        opts['<format>'] = 'json'
    else:
        _die_usage('Unknown extension: {0}'.format(ext))

    return opts


def main():
    opts = _parse_args()
    last_names, female_first_names, male_first_names = _load_data()
    people = []
    salaries = None
    total = opts['<total>']
    if opts['--salary']:
        (low, high) = opts['--salary']
        salaries = _generate_salaries(total, low, high)

    total_females = int(total * opts['--female'] / 100)
    total_males   = int(total * opts['--male'] / 100)
    delta = abs(total - total_females - total_males)
    total_females += delta

    # Generate the invariant fields.
    for first_names, n, gender in ((female_first_names, total_females, 'F'),
                                   (male_first_names, total_males, 'M')):
      for i in range(n):
          person = {}
          first, middle, last = _generate_name(first_names, last_names)
          person['firstName'] = first
          person['middleName'] = middle
          person['lastName'] = last
          person['gender'] = gender
          person['birthDate'] = _generate_birth_date(opts['--yearmin'],
                                                     opts['--yearmax'])
          people.append(person)

    # Generate everything else
    for i, person in enumerate(people):
        if opts['--ssn']:
            person['ssn'] = _make_fake_ssn()

        if opts['--salary']:
            person['salary'] = salaries[i]

    if opts['<format>'] == 'json':
        _write_json(opts['<outputfile>'], people)
    else:
        _write_csv(file=opts['<outputfile>'],
                   people=people,
                   delim=opts['--sep'],
                   header=opts['--header'],
                   salaries=opts['--salary'] is not None,
                   ssn=opts['--ssn'])

if __name__ == '__main__':
    main()
